## 第6章　生成

In [46]:
import numpy as np
import pandas as pd

In [34]:
from imblearn.over_sampling import SMOTE

### 6-2　オーバーサンプリングによる不均衡データの調整

In [35]:
sm = SMOTE(
    
    # 少ないデータを、多い方のデータの何割まで増やすか
    sampling_strategy='auto',
    
    # kパラメータ
    k_neighbors=5, random_state=71)

In [36]:
production_tb = pd.read_csv('production.csv')

print(production_tb.shape)
production_tb.head()

(1000, 4)


type      length  thickness  fault_flg
0    E  274.027383  40.241131      False
1    D   86.319269  16.906715      False
2    E  123.940388   1.018462      False
3    B  175.554886  16.414924      False
4    B  244.934740  29.061081      False

In [37]:
print('True:',len(production_tb[production_tb['fault_flg'] == True]))
print('False:',len(production_tb[production_tb['fault_flg'] == False]))

True: 52
False: 948


障害なし（fault_flgがFalse）が948件に対して、障害なし（fault_flgがTrue）が52件と少ないので、これを解消する。

In [38]:
blance_data, blance_target = sm.fit_sample(production_tb[['length', 'thickness']], production_tb['fault_flg'])

In [39]:
blance_data.head()

length  thickness
0  274.027383  40.241131
1   86.319269  16.906715
2  123.940388   1.018462
3  175.554886  16.414924
4  244.934740  29.061081

In [40]:
blance_target.head()

0    False
1    False
2    False
3    False
4    False
Name: fault_flg, dtype: bool

In [57]:
# 同じ数になっている
print(len(blance_target[blance_target == True]))
print(len(blance_target[blance_target == False]))

948
948
